Tutorial (#35)
mlabs-haskell/ytxp-lib: ytxp-plutarch/examples/direct-offer/doc/tutorials/specification.md

## How to write the direct offer specification

The specification generally includes one component specification if using the component model, along with one or more transaction family specifications. Additionally, it may contain one or more transaction flow specifications for more complex cases.

### Component

For this contract, we will utilize the component model. In this model, an offer order is represented by a UTxO that includes a datum detailing the offer's specifics, and a state thread token verifies the semantic validity of the offer UTxO.

For the specification, we will detail the existence of a component and describe all its attributes, including the UTxO's address, value, datum, and reference script.

Specifically, we will define the attributes of the offer component in the following way:

- We will not specify any particular values, as the component's value is what the consumer can keep.
- The address must be a YTxP yielding-validator address to ensure that the yielding validator will only validate when an authorized transaction family is also validated.

> - The datum consists of two fields: the owner and the to buy value.

Reviewer: ```suggestion
- The datum consists of two fields: the "owner" and the "to buy" value.
```

- No scripts are allowed to be referenced.

If a component has multiple states, we will specify that in the documentation. However, the offer component has no meaningful states, so we will not provide any details.

> Finally, we will specify which transaction family will introduce, modify, and terminate this component. Please visit to the example of the [offer component specification](/ytxp-plutarch/examples/direct-offer/doc/components/offer.md).

Reviewer: ```suggestion
Finally, we will specify which transaction family will introduce, modify, and terminate this component. Have a look the example of the [offer component specification](/ytxp-plutarch/examples/direct-offer/doc/components/offer.md) for more details.
```

> The component specification serves only as information and does not detail the transaction validation for the offer component. Validation of the component is delegated to the authorized transaction families.

Reviewer: ```suggestion
The component specification does not detail the transaction validation for the offer component. Validation of the component is delegated to the authorised transaction families.
```

### Transaction families
To determine the transaction families involved in the direct offer contract, we need to identify all atomic transactions that can take place.

- Create an order
- Reclaim an order
- Execute an order

These transactions represent the transaction families: Creating, Reclaiming, and Executing.

A transaction family specification outlines the details necessary for transaction validation.

> The transaction family itself is an authorized script that can be implemented with any validator types. However, it is recommended to use the staking validator, as it only requires a zero withdrawal to activate this validator.

Reviewer: ```suggestion
The transaction family itself is an authorised script that can be implemented with any validator types. However, it is recommended to use the staking validator, as it only requires a zero withdrawal to activate this validator.
```
